go-controller/pkg/crd/clusteruserdefinednetwork/v1/types.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	udnv1 "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/userdefinednetwork/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ClusterUserDefinedNetwork describe network request for a shared network across namespaces.
+//
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:path=clusteruserdefinednetworks,scope=Cluster
+// +kubebuilder:singular=clusteruserdefinednetwork
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+type ClusterUserDefinedNetwork struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// +kubebuilder:validation:Required
+	// +required
+	Spec ClusterUserDefinedNetworkSpec `json:"spec"`
+	// +optional
+	Status ClusterUserDefinedNetworkStatus `json:"status,omitempty"`
+}
+
+// ClusterUserDefinedNetworkSpec defines the desired state of ClusterUserDefinedNetwork.
+type ClusterUserDefinedNetworkSpec struct {
+	// NamespaceSelector Label selector for which namespace network should be available for.
+	// +kubebuilder:validation:Required
+	// +required
+	NamespaceSelector metav1.LabelSelector `json:"namespaceSelector"`
+
+	// Template is direct specification of UserDefinedNetwork.
+	// +kubebuilder:validation:Required
+	// +required
+	Template UserDefinedNetworkTemplateSpec `json:"template"`
+}
+
+// UserDefinedNetworkTemplateSpec UserDefinedNetwork spec template.
+type UserDefinedNetworkTemplateSpec struct {
+	// UserDefinedNetworkSpec contains the UserDefinedNetwork specification.
+	Spec udnv1.UserDefinedNetworkSpec `json:"spec,omitempty"`
```

Comment:
You right, the Cluster UDN CRD should have its own spec as it may need to diverge form the UDN CRD spec.
I will change it accordingly.
 
Regarding localnet, yes we did it fit better the cluster UDN CRD as it targeted for cluster admin only.
Although it seems exposing localnet require more thinking as it may not fit with future changed for localnet implementation such as https://github.com/ovn-org/ovn-kubernetes/pull/4320
I would rather add support for localnet on a follow up PR, can we do that?

Reply:
> I would rather add support for localnet on a follow up PR, can we do that?

I think it's fine, since that's not in scope for primary UDN. Whatever's not about primary UDN can be left for another release.